drivers/sensor/pmw3360/pmw3360.c

```diff
@@ -0,0 +1,938 @@
+	}
+
+	LOG_DBG("Optical chip firmware ID: 0x%x", fw_id);
+	if (fw_id != PMW3360_FIRMWARE_ID) {
+		LOG_ERR("Chip is not running from SROM!");
+		return -EIO;
+	}
+
+	u8_t product_id;
+	err = reg_read(dev_data, PMW3360_REG_PRODUCT_ID, &product_id);
+	if (err) {
+		LOG_ERR("Cannot obtain product id");
+		return err;
+	}
+
+	if (product_id != PMW3360_PRODUCT_ID) {
+		LOG_ERR("Invalid product id!");
+		return -EIO;
+	}
+
+	/* Write 0x20 to Config2 register for wireless mouse design.
+	 * This enables entering rest modes.
+	 */
+	err = reg_write(dev_data, PMW3360_REG_CONFIG2, 0x20);
+	if (err) {
+		LOG_ERR("Cannot enable REST modes");
+	}
+
+	return err;
+}
+
+static void irq_handler(struct device *gpiob, struct gpio_callback *cb,
+			u32_t pins)
+{
+	sensor_trigger_handler_t handler;
+
+	k_spinlock_key_t key = k_spin_lock(&pmw3360_data.lock);
+	handler = pmw3360_data.data_ready_handler;
+	k_spin_unlock(&pmw3360_data.lock, key);
+
+	struct sensor_trigger trig = {
+		.type = SENSOR_TRIG_DATA_READY,
+		.chan = SENSOR_CHAN_ALL,
+	};
+	handler(DEVICE_GET(pmw3360), &trig);
+}
+
+static int pmw3360_async_init_power_up(struct pmw3360_data *dev_data)
+{
+	/* Reset sensor */
+
+	spi_cs_ctrl(dev_data, false);
+
+	return reg_write(dev_data, PMW3360_REG_POWER_UP_RESET, 0x5A);
+}
+
+static int pmw3360_async_init_configure(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	err = update_cpi(dev_data, CONFIG_PMW3360_CPI);
+
+	if (!err) {
+		err = update_downshift_time(dev_data,
+					    PMW3360_REG_RUN_DOWNSHIFT,
+					    CONFIG_PMW3360_RUN_DOWNSHIFT_TIME_MS);
+	}
+
+	if (!err) {
+		err = update_downshift_time(dev_data,
+					    PMW3360_REG_REST1_DOWNSHIFT,
+					    CONFIG_PMW3360_REST1_DOWNSHIFT_TIME_MS);
+	}
+
+	if (!err) {
+		err = update_downshift_time(dev_data,
+					    PMW3360_REG_REST2_DOWNSHIFT,
+					    CONFIG_PMW3360_REST2_DOWNSHIFT_TIME_MS);
+	}
+
+	if (!err) {
+		dev_data->ready = true;
+	}
+
+	return err;
+}
+
+static void pmw3360_async_init(struct k_work *work)
+{
+	struct pmw3360_data *dev_data = &pmw3360_data;
+
+	ARG_UNUSED(work);
+
+	LOG_DBG("PMW3360 async init step %d", dev_data->async_init_step);
+
+	dev_data->err = async_init_fn[dev_data->async_init_step](dev_data);
+	if (dev_data->err) {
+		LOG_ERR("PMW3360 initailization failed");
+	} else {
+		dev_data->async_init_step++;
+
+		if (dev_data->async_init_step == ASYNC_INIT_STEP_COUNT) {
+			dev_data->ready = true;
+			LOG_INF("PMW3360 initialized");
+		} else {
+			k_delayed_work_submit(&dev_data->init_work,
+					      async_init_delay[dev_data->async_init_step]);
+		}
+	}
+}
+
+static int pmw3360_init_cs(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	dev_data->cs_gpio_dev =
+		device_get_binding(CONFIG_PMW3360_CS_GPIO_DEV_NAME);
+	if (!dev_data->cs_gpio_dev) {
+		LOG_ERR("Cannot get CS GPIO device");
+		return -ENXIO;
+	}
+
+	err = gpio_pin_configure(dev_data->cs_gpio_dev,
+				 CONFIG_PMW3360_CS_GPIO_PIN,
+				 GPIO_DIR_OUT);
+	if (err) {
+		LOG_ERR("Cannot configure CS PIN");
+	}
+
```

Comment:
The CS pin should be driven to the initial (inactive) state here. With the current GPIO API it is not possible to do it in one shot while the pin is configured, but this is to change soon.

Reply:
done